evaluator/builtin_other.go

```diff
@@ -0,0 +1,327 @@
+	return func(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+		a, b := types.CoerceDatum(args[0], args[1])
+
+		if a.Kind() == types.KindNull || b.Kind() == types.KindNull {
+			d.SetNull()
+			return
+		}
+
+		x, err := a.ToInt64()
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		y, err := b.ToInt64()
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		// use a int64 for bit operator, return uint64
+		switch op {
+		case opcode.And:
+			d.SetUint64(uint64(x & y))
+		case opcode.Or:
+			d.SetUint64(uint64(x | y))
+		case opcode.Xor:
+			d.SetUint64(uint64(x ^ y))
+		case opcode.RightShift:
+			d.SetUint64(uint64(x) >> uint64(y))
+		case opcode.LeftShift:
+			d.SetUint64(uint64(x) << uint64(y))
+		default:
+			err = ErrInvalidOperation.Gen("invalid op %v in bit operation", op)
+			return
+		}
+		return
+	}
+}
+
+func arithmeticFuncFactory(op opcode.Op) BuiltinFunc {
+	return func(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+		a, err := coerceArithmetic(args[0])
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		b, err := coerceArithmetic(args[1])
+		if err != nil {
+			err = errors.Trace(err)
+			return
+		}
+
+		a, b = types.CoerceDatum(a, b)
+		if a.Kind() == types.KindNull || b.Kind() == types.KindNull {
+			d.SetNull()
+			return
+		}
+
+		switch op {
+		case opcode.Plus:
+			return computePlus(a, b)
+		case opcode.Minus:
+			return computeMinus(a, b)
+		case opcode.Mul:
+			return computeMul(a, b)
+		case opcode.Div:
+			return computeDiv(a, b)
+		case opcode.Mod:
+			return computeMod(a, b)
+		case opcode.IntDiv:
+			return computeIntDiv(a, b)
+		default:
+			err = ErrInvalidOperation.Gen("invalid op %v in arithmetic operation", op)
+			return
+		}
+	}
+}
+
+func unaryOpFactory(op opcode.Op) BuiltinFunc {
+	return func(args []types.Datum, _ context.Context) (d types.Datum, err error) {
+		defer func() {
+			if er := recover(); er != nil {
+				err = errors.Errorf("%v", er)
+			}
+		}()
+		aDatum := args[0]
+		if aDatum.Kind() == types.KindNull {
+			d.SetNull()
+			return
+		}
+		switch op {
+		case opcode.Not:
+			var n int64
+			n, err = aDatum.ToBool()
+			if err != nil {
+				err = errors.Trace(err)
+			} else if n == 0 {
+				d.SetInt64(1)
+			} else {
+				d.SetInt64(0)
+			}
+		case opcode.BitNeg:
+			var n int64
+			// for bit operation, we will use int64 first, then return uint64
+			n, err = aDatum.ToInt64()
+			if err != nil {
+				err = errors.Trace(err)
+				return
+			}
+			d.SetUint64(uint64(^n))
+		case opcode.Plus:
+			switch aDatum.Kind() {
+			case types.KindInt64,
+				types.KindUint64,
+				types.KindFloat64,
+				types.KindFloat32,
+				types.KindMysqlDuration,
+				types.KindMysqlTime,
+				types.KindString,
+				types.KindMysqlDecimal,
+				types.KindBytes,
+				types.KindMysqlHex,
+				types.KindMysqlBit,
+				types.KindMysqlEnum,
+				types.KindMysqlSet:
+				d = aDatum
+			default:
+				err = ErrInvalidOperation
```

Comment:
Add some error info, for example ErrInvalidOperation.Gen("Unsupported type %v for opcode.Plus", aDatum.Kind())